src/System.Management.Automation/CoreCLR/CorePsAssemblyLoadContext.cs

```diff
@@ -240,7 +240,26 @@ internal static IntPtr NativeDllHandler(Assembly assembly, string libraryName)
         {
             s_nativeDllSubFolder ??= GetNativeDllSubFolderName(out s_nativeDllExtension);
             string folder = Path.GetDirectoryName(assembly.Location);
-            string fullName = Path.Combine(folder, s_nativeDllSubFolder, libraryName) + s_nativeDllExtension;
+
+            // Some developers specify `.dll` in their DllImport attribute even though
+            // they ship native libraries for Linux and macOS, but they aren't found because the extension is now
+            // treated as part of the filename. If the libraryName contains a known extension for wrong OS, we
+            // remove it and add the OS specific extension.
+            var extension = Path.GetExtension(libraryName).ToLowerInvariant();
+            string fullName;
+
+            if (string.IsNullOrEmpty(extension))
+            {
+                fullName = Path.Combine(folder, s_nativeDllSubFolder, libraryName) + s_nativeDllExtension;
+            }
+            else if (extension != s_nativeDllExtension)
+            {
+                fullName = Path.Combine(folder, s_nativeDllSubFolder, Path.GetFileNameWithoutExtension(libraryName)) + s_nativeDllExtension;
```

Comment:
This could be controversial. Is it worth changing an extension that is specified by the developer in his code?

Reply:
Please look https://github.com/PowerShell/PowerShell/issues/20740#issuecomment-1854616098

Reply:
Isn't this the problem with the sqlite repro?  It seems that the developer accidentally specified .dll but also shipped libraries for other OS.

Reply:
There is no requirement that a shared library does end with dylib, dll or so. merely convention. If you remember on Windows control panel extensions, drivers and OLE controls were dlls but did not have dll as the extension.

I don't think the developer accidentally specified .dll in the DllImport, the examples from Microsoft have "user32.dll" etc in them. As the module is common and only the shared libraries are different, it is consistent to have them all with the same name but just in a different directory according to architecture.

Reply:
[DllImport documentation](https://learn.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.dllimportattribute.value?view=net-8.0) says that we can use `You can provide a full or relative path.`  or [`As a minimum requirement, you must supply the name of the DLL containing the entry point.`](https://learn.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.dllimportattribute?view=net-8.0) and says nothing about extensions. So developers are free to use any file extension for dynamic library.
While I agree that deviating from the usual values looks amazing, there are no standards to call it a bug, especially since these non-standard extensions work in C# applications. If we forcibly change the extension, then obviously these applications will stop working.

Reply:
Perhaps if it has an extension, we try loading it as-is, and if it fails, then try the os specific extension?

Reply:
I suggest not removing anything from the string that was originally given, trust that the original string was correct, because if it is in published code it must have been tested like that and worked.

The simplest would be...

1. try the exact given name

2. if that does not work try with operating system specific extension appended

No replacing of the extension with something else, only appending.

So if you don't specify an extension you will get the OS one. 

If you do, your extension will be honoured first, then it will try with your full name plus OS extension.

And of course it has to go through (a) in the root directory (b) in the directory with just the CPU name (c) in the directory with OS-CPU

So may take 6 attempts to load the DLL.

Reply:
The fallback doesn't make a lot of sense. If a developer specified `DllImport("interop.msft")` in the code, then obviously he is referring to the real `interop.msft` file on file system and there is no `interop.msft.dll` file or `interop.dll` file.
(And I don't think DllImport is smart enough to handle multi-dot names. I mean `DllImport("interop.msft")` will do not search `interop.msft.dll` on Windows and `interop.msft.so` on Linux.)